### PR TITLE
support querying multiple patterns within a single query by introduci…

### PR DESCRIPTION
**Pull request by swilly22**
…ng join operation

### CODE REVIEW

File: src/execution_plan/execution_plan.c [hunk 1]
```diff
@@ -248,40 +248,83 @@ ExecutionPlan* NewExecutionPlan(RedisModuleCtx *ctx, Graph *g, const char *graph
...
+
+        // For every pattern in match clause.
+        size_t patternCount = Vector_Size(ast->matchNode->patterns);
+        
+        /* Incase we're dealing with multiple patterns
+         * we'll simply join them all together with a join operation. */
+        bool multiPattern = patternCount > 1;
+        OpNode *cartesianProduct = NULL;
+        if(multiPattern) {
+            cartesianProduct = NewOpNode(NewCartesianProductOp());
```
Comment (jeffreylovitz) [inline src/execution_plan/execution_plan.c]: There's no way around it, of course, but I'm scared just thinking of the cost of a Cartesian product with _n_ operands! Ick.

File: src/execution_plan/execution_plan.c [hunk 2]
```diff
@@ -248,40 +248,83 @@ ExecutionPlan* NewExecutionPlan(RedisModuleCtx *ctx, Graph *g, const char *graph
...
-                    op = NewOpNode(NewCondTraverseOp(g, q, exps[i]));
-                    Vector_Push(ops, op);
+            if(multiPattern) {
+                // Connect cartesian product to the root of traversal.
+                Vector_Get(traversals, 0, &op);
+                _OpNode_AddChild(cartesianProduct, op);
+
+                // Connect traversal operations.
+                OpNode *childOp;
+                OpNode *patentOp;
```
Comment (jeffreylovitz) [inline src/execution_plan/execution_plan.c]: Typo - should be `parentOp`.

File: src/graph/query_graph.c [hunk 3]
```diff
@@ -101,14 +101,37 @@ void _BuildQueryGraphAddProps(AST_GraphEntity *entity, GraphEntity* e) {
...
-    QueryGraph_AddNode(graph, n, entity->alias);
+    Node *n = QueryGraph_GetNodeByAlias(graph, entity->alias);
+    if(n == NULL) {
+        /* Create a new node, set its properties, and add it to the graph. */
+        Node *n = Node_New(INVALID_ENTITY_ID, entity->label);
+        _BuildQueryGraphAddProps(entity, (GraphEntity*)n);
+        QueryGraph_AddNode(graph, n, entity->alias);
+    } else {
+        /* Merge nodes. */
+        _MergeNodeWithGraphEntity(n, entity);
```
Comment (jeffreylovitz) [inline src/graph/query_graph.c]: I'm not sure that I understand this. If we refer to the same node multiple times in a query using different aliases, shouldn't it be treated as multiple distinct entities?

File: tests/flow/test_multi_pattern.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        global redis_graph
+        query = """MATCH(r:person), (f:person) CREATE (r)-[:friend]->(f)"""
+        actual_result = redis_graph.query(query)
+        assert (actual_result.relationships_created == 49)
+    
+    def test_04_verify_fully_connected_graph(self):
+        global redis_graph
+        query = """MATCH(r:person)-[]->(f:person) RETURN count(r)"""
+        actual_result = redis_graph.query(query)
+        friend_count = int(float(actual_result.result_set[1][0]))
```
Comment (jeffreylovitz) [inline tests/flow/test_multi_pattern.py]: Can't we cast directly to int?

Comment (swilly22) [reply tests/flow/test_multi_pattern.py]: Nope, as numerics are doubles, redis returns a string representation of a number with a floating point, as a result we must cast to float first.

**Pull request merged**